awscli/customizations/s3/results.py

```diff
@@ -121,3 +122,77 @@ def _get_src_dest(self, future):
         src = 's3://' + copy_source['Bucket'] + '/' + copy_source['Key']
         dest = 's3://' + call_args.bucket + '/' + call_args.key
         return src, dest
+
+
+class ResultRecorder(object):
+    """Records and track transfer statistics based on results receieved"""
+    def __init__(self):
+        self.bytes_transferred = 0
+        self.bytes_failed_to_transfer = 0
+        self.files_transferred = 0
+        self.files_failed = 0
+        self.files_warned = 0
+        self.expected_bytes_transferred = 0
+        self.expected_files_transferred = 0
+
+        # Keys: result from _get_progress_cache_key()
+        # Values: number of bytes left for a particular transfer
+        self._progress_remaining_cache = {}
+
+        self._result_handler_map = {
+            QueuedResult: self._record_queued_result,
+            ProgressResult: self._record_progress_result,
+            SuccessResult: self._record_success_result,
+            FailureResult: self._record_failure_result,
+            WarningResult: self._record_warning_result,
+        }
+
+    def record_result(self, result):
+        """Record the result of an individual Result object"""
+        self._result_handler_map.get(type(result), self._record_noop)(
+            result=result)
+
+    def _get_progress_cache_key(self, result):
+        return ':'.join([result.transfer_type, result.src, result.dest])
```

Comment:
That seems reasonable.  Even if a type check was added, it would still likely be an error case anyways, and the main benefit there would just be a better error message than `AttributeError`.